[Merged by Bors] - feat(analysis/special_functions/trigonometric): range_{exp,cos,sin}

jcommelin opened the pull request:
---

- [x] depends on: #4656 

<!--
put comments you want to keep out of the PR commit here.
If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

jcommelin committed:
feat(analysis/special_functions/trigonometric): range_cos_infinite

ADedecker commented on src/analysis/special_functions/trigonometric.lean:
```diff
@@ -1754,6 +1755,24 @@ lemma continuous_tan : continuous (λ x : {x | cos x ≠ 0}, tan x) :=
+  { rintro ⟨x, hx⟩ ⟨y, hy⟩ h,
+    simpa only [complex.of_real_inj, subtype.mk_eq_mk] using h }
+end
+
```
```suggestion
lemma range_cos_infinite : (set.range complex.cos).infinite :=
begin
  have : (set.range real.cos).infinite,
  { rw real.range_cos,
    apply Icc.infinite,
    norm_num },
  delta real.cos at this,
  rw [range_comp, range_comp] at this,
  exact set.infinite_mono
    (image_subset_range _ (range (coe : ℝ → ℂ))) (λ h, this (set.finite.image _ h)),
end
```
Having to use injectivity felt weird, so I wanted to try another way. Maybe that means we should have a lemma saying `(f '' s).infinite -> s.infinite` so that there's no need to go back to `set.finite.image`

jcommelin replied:
Thanks!

jcommelin committed:
Update src/analysis/special_functions/trigonometric.lean

Co-authored-by: Anatole Dedecker <anatolededecker@gmail.com>

jcommelin committed:
factor out sublemmas; copy for sin

urkud commented on src/analysis/special_functions/trigonometric.lean:
```diff
@@ -1754,8 +1761,23 @@ lemma continuous_tan : continuous (λ x : {x | cos x ≠ 0}, tan x) :=
 by { rw continuous_on_iff_continuous_restrict, convert continuous_tan }
 
-end complex
+lemma range_cos_infinite : (range complex.cos).infinite :=
```
Should we add `real.range_exp`, `complex.range_exp`, `complex.range_cos` and `complex.range_sin`?

jcommelin replied:
I've added those lemmas.

jcommelin committed:
wip

jcommelin committed:
refactor(algebra/quadratic_discriminant): drop linearity condition; cleanup

Renames:

- `discriminant_le_zero` to `discrim_le_zero`
- `discriminant_lt_zero` to `discrim_lt_zero`

jcommelin committed:
weaken from lin.ord.fld to lin.ord.semiring

jcommelin committed:
finish proofs

jcommelin committed:
generalise

urkud commented on src/analysis/special_functions/trigonometric.lean:
```diff
@@ -1658,6 +1669,21 @@ lemma log_I : log I = π / 2 * I := by simp [log]
 
 lemma log_neg_I : log (-I) = -(π / 2) * I := by simp [log]
 
+lemma exists_pow_nat_eq (x : ℂ) {n : ℕ} (hn : 0 < n) : ∃ z, z ^ n = x :=
```
TODO (not necessarily for this PR): generalize these two lemmas to an `[is_alg_closed]` field and add an instance `[is_alg_closed complex]`.

bryangingechen replied:
It would be a good idea to create an issue if you won't be working on it immediately.

bryangingechen replied:
Maybe I'm misunderstanding the comment, but we did have the `complex.is_alg_closed` instance: https://github.com/leanprover-community/mathlib/blob/09afb0465df681c12c2e53d8fb88e7c6ff7e807f/src/field_theory/algebraic_closure.lean#L72

urkud commented on src/analysis/special_functions/trigonometric.lean:
```diff
@@ -1862,8 +1888,56 @@ lemma continuous_tan : continuous (λ x : {x | cos x ≠ 0}, tan x) :=
+@[simp] lemma range_cos : range cos = set.univ :=
+cos_surjective.range_eq
+
+lemma sin_surjective : function.surjective sin :=
```
You can use `sin_add_pi_div_two` here.

PatrickMassot committed:
simplify surjective_sin following Yury

PatrickMassot committed:
Typo
